imxrt-hal/src/tempmon.rs

```diff
@@ -0,0 +1,395 @@
+        t.set_measure_frequency(measure_freq);
+        t
+    }
+}
+
+/// A Temperature Monitor (TEMPMON)
+///
+/// # Example 1
+///
+/// ```no_run
+/// use imxrt1060_hal;
+///
+/// let mut peripherals = imxrt1060_hal::Peripherals::take().unwrap();
+/// let (_, ipg_hz) = peripherals.ccm.pll1.set_arm_clock(
+///     imxrt1060_hal::ccm::PLL1::ARM_HZ,
+///     &mut peripherals.ccm.handle,
+///     &mut peripherals.dcdc,
+/// );
+/// let mut cfg = peripherals.ccm.perclk.configure(
+///     &mut peripherals.ccm.handle,
+///     imxrt1060_hal::ccm::perclk::PODF::DIVIDE_3,
+///     imxrt1060_hal::ccm::perclk::CLKSEL::IPG(ipg_hz),
+/// );
+///
+/// // init temperature monitor
+/// // consider using init_with_measure_freq
+/// let mut temp_mon = peripherals.tempmon.init();
+/// loop {
+///     if let Ok(_temperature) = nb::block!(temp_mon.measure_temp()) {
+///         // _temperature in mC (1°C = 1000°mC)
+///     }
+/// }
+/// ```
+pub struct TempMon {
+    base: ral::tempmon::Instance,
+    /// scaler * 1000
+    scaler: i32,
+    /// hot_count * 1000
+    hot_count: i32,
+    /// hot_temp * 1000
+    hot_temp: i32,
+
+    /// measurement active flag to get measure_temp unblocking
+    measurement_active: bool,
+}
+
+impl TempMon {
+    /// converts the temp_cnt into a human readable temperature [°mC] (1/1000 °C)
+    fn convert(&self, temp_cnt: i32) -> i32 {
+        let n_meas = temp_cnt - self.hot_count;
+        self.hot_temp - n_meas * self.scaler
+    }
+
+    /// decode the temp_value into measurable bytes
+    ///
+    /// param **temp_value_mc**: in °mC (1/1000°C)
+    ///
+    fn decode(&self, temp_value_mc: i32) -> u32 {
+        let v = (temp_value_mc - self.hot_temp) / self.scaler;
+        (self.hot_count - v) as u32
+    }
+
+    /// triggers a new measurement and waits until it's finished
+    ///
+    /// If you configured automatically repeating, this will trigger additional measurement.
+    /// Use get_temp instate to get the last read value
+    ///
+    /// The returning temperature in 1/1000 Celsius (°mC)
+    ///
+    /// Example: 25500°mC -> 25.5°C
+    pub fn measure_temp(&mut self) -> nb::Result<i32, TempMonError> {
+        if !self.is_powered_up() {
+            Err(nb::Error::from(TempMonError::PowerDown))
+        } else {
+            // if no measurement is active, trigger new measurement
+            if !self.measurement_active {
+                ral::write_reg!(ral::tempmon, self.base, TEMPSENSE0, MEASURE_TEMP: 1);
+                self.measurement_active = true;
+            }
+
+            // if the measurement is not finished or not started
+            // i.MX Docs: This bit should be cleared by the sensor after the start of each measurement
+            if ral::read_reg!(ral::tempmon, self.base, TEMPSENSE0, FINISHED == 0) {
+                // this could be triggered again without any effect
+                Err(nb::Error::WouldBlock)
+            } else {
+                self.measurement_active = false;
```

Comment:
Is it important to maintain this state in RAM? This seems like information we already have in the `MEASURE_TEMP` field. Instead of setting this bool to `false`, we could set `MEASURE_TEMP = 0`.

Reply:
believe it or not, I double-checked it and I didn't know why I assumed this behavior. 

Probably I expected this that way because I start the next measurement with a 1. 
I learned a lot about memory-mapping and how it works. Thanks for your patience :-)

Reply:
No problem! Thanks for double checking, and for updating the approach. It wouldn't have surprised me if the behavior was undocumented / missing in the reference manual 😛.